### PR TITLE
Exit after compilation when syntax errors occurred (closes #280)

### LINKED ISSUE #280: Inconsistent handling of unexpected symbol syntax errors
This is an issue that appeared while debugging issue #277 . I copied the text out of the [related issue comment](https://github.com/cksystemsteaching/selfie/issues/277#issuecomment-831565963).

Selfie does not terminate with a non-zero return code to indicate compilation issues. Unexpected symbol handling seems somewhat inconsistent as sometimes Selfie does exit with a non-zero code immediately, and some other times Selfie does print a message, but continues compilation.
https://github.com/cksystemsteaching/selfie/blob/e113d85a477779c9d545ff68b31fc1b3b72f8c66/selfie.c#L4435-L4442
https://github.com/cksystemsteaching/selfie/blob/e113d85a477779c9d545ff68b31fc1b3b72f8c66/selfie.c#L4546-L4559

---------------------

[Follow-up comment from @ckirsch](https://github.com/cksystemsteaching/selfie/issues/277#issuecomment-831725714) : 
> @Blazefrost Interesting! Does gcc return a non-zero exit code if there were warnings but no errors? In that case we may want to mimic that behavior.

### PR DESCRIPTION
**Pull request by Blazefrost**
This pull request introduces the EXITCODE_SYNTAXERROR (and also adds a missing exit code to the grader).

Previously, Selfie exited on non-recoverable syntax errors (e.g. EXITCODE_SCANNERERROR for unknown escape sequences, EXITCODE_PARSERERROR for undeclared variables or EXITCODE_COMPILERERROR if there are no temporary registers left). However, it silently continued on recoverable syntax errors, treating them more like warnings.

This commit introduces the counter `number_of_syntax_errors` which shall be incremented each time a syntax error was encountered. After compilation, Selfie checks the syntax error counter and exits with a message before the binary is written to the disk if there actually were syntax.
Because compilation is still not immediately stopped after encountering a recoverable syntax error, it is still possible to enumerate all syntax errors of a file.

### CODE REVIEW

**Review by ckirsch — CHANGES_REQUESTED** [review]
@Blazefrost Thanks a lot! See my comments on two tiny issues.

**Review by ckirsch — CHANGES_REQUESTED** [review]

File: grader/assignments/fork-wait/sum-exit-code.c [hunk 1]
```diff
@@ -86,7 +86,7 @@ int main(int argc, char** argv) {
   // prepare data to be printed
   i = 0;
   while (i < number_of_processes) {
-    *(sorted_numbers + i) = i;
+    *(sorted_numbers + i) = i*2;
```
[the remark this answers is not in the document]
Comment (Blazefrost) [reply grader/assignments/fork-wait/sum-exit-code.c:89]: I've explained the changes on the grader more detailed in commit 463762a.

In summary, the grader's test suite makes sure that no exit codes of any checks collide with any of Selfie's exit codes. By adding `EXITCODE_SYNTAXERROR` (and another exit code that has not been added to the grader yet), the old return value of the `sum-exit-code` would collide with `EXITCODE_MODELINGERROR`. I've fixed this by doubling the exit code.

File: selfie.c [hunk 2]
```diff
@@ -451,6 +451,8 @@ uint64_t number_of_ignored_characters = 0;
 uint64_t number_of_comments           = 0;
 uint64_t number_of_scanned_symbols    = 0;
 
+uint64_t number_of_syntax_errors = 0; // the number of encountered syntax error
```
[the remark this answers is not in the document]
Comment (Blazefrost) [reply selfie.c]: Done!

File: selfie.c [hunk 3]
```diff
@@ -5740,6 +5759,7 @@ void selfie_compile() {
         exit(EXITCODE_IOERROR);
       }
 
+      number_of_syntax_errors = 0;
```
[the remark this answers is not in the document]
Comment (Blazefrost) [reply selfie.c]: Done!

**Comment by ckirsch** [comment]
@Blazefrost I merged your code into the branch `elementary-computer-science`. Could you please check if your changes here work there as well? In particular, I moved the exit codes for symbolic execution and modeling into the engine and modeler...

**Comment by Blazefrost** [comment]
I tested it on `elementary-computer-science` and it does work fine!

**Pull request merged**